### DISCUSSION TITLE
How to listen data in stream, while write it at the same time

### DISCUSSION THREAD

**Discussion by lucasjinreal** [simolus3/drift#2937] [Q&A]
Hi, I have a code like this:

```
Stream<List<types.Message>> getUiMessagesForRoom(ContactChat ctc, int pageNum,
      {int perPage = 32}) {
    Stream<List<DbMessageWithDbContactOrRoomMember>> a = AppData.instance()!
        .getMsgsForRoom(ctc, pageNum: pageNum, perPage: perPage);
    var b = a.map((rawData) {
      setMsgDisplayed(rawData);
}
}
```

It basically reads data from drift db, which returns a stream, my data needs to update in real time, so I listen it as stream to update it.

But I need to write it when query it, but this got me crazy: when write it will update, caused a little cycle problem.

Does there any elegant way to deal with this situation/.

**Comment by simolus3** [comment]
There are always going to be a few cycles of stream emissions -> updates -> emissions with these setups, but it should be possible to at least not make them endless.
Does `setMsgDisplayed` always do a write to the database? If it only runs an update if the message is not already marked as read, the cycle should eventually come to an end as all messages have been marked as read. Using `distinct()` on the stream to avoid getting notifications after a write which didn't change the returned rows could also work.

**Reply by lucasjinreal** [reply]
Hi, setMsgDisplayed will runs untill all messages marked as read.

But this looks like could run many times therotically even thought in can stream listen func print didn't call endless times, I think this might because of when all messages marked as read, there wont't be more emitions.

Does the `distinct` could be: return the messages only, when database new added but not updated existed one?
